### ISSUE TITLE
[java] StringIndexOutOfBoundsException

### ISSUE THREAD

**Issue by ilyash-b** [jcoglan/canopy#23]
Hi.

Offending generated code:
```
            if (offset < inputSize) {
                chunk0 = input.substring(offset, offset + 4);
            }
```
The code above throws StringIndexOutOfBoundsException when the parser is towards the end of input.

I guess should be `if (offset + 4 < ...) ...`

This happens when trying to match a literal string in the grammar.

Might help:
```
groovy:000> "abc".substring(1,2)
===> b
groovy:000> "abc".substring(1,20)
ERROR java.lang.StringIndexOutOfBoundsException:
String index out of range: 20
        at java_lang_String$substring$2.call (Unknown Source)
```

**Comment by avasenin** [comment]
This bug was fixed by https://github.com/jcoglan/canopy/commit/c5d4f74922ebf2f3f81c92fe429bf7c855388770 but not released to npm.

@jcoglan can you please release the new version to npm?

**Comment by ilyash-b** [comment]
Hi.

First, thanks!
Second - unfortunately I will not have time to test it in the near future, we (at work) have decided to prioritize other things so Canopy will not be used for now.
Third - (following from second) I don’t see tests in the fixing commit. I can not and I'm not telling you what to do but … how do you know the fix works?

**Comment by fireboy1919** [comment]
I can chime in that I had the same problem, and I fixed it by switching to master (which looks like it's the same, except with two extra bugfixes).

I'd like to use this for java, but obviously this particular bug is a showstopper.
Are you proposing that if we want this in npm we should write a java test suite, or are you just looking for verbal confirmation that it works?

**Comment by linehammer** [comment]
When using the Java substring() method, a subset of the character sequence can be extracted from a string. The [substring](http://net-informations.com/java/string/substring.htm) index must be any value from 0 to the length of a string. The java.lang.StringIndexOutOfBoundsException thrown by String methods to indicate that the beginIndex is negative , or endIndex is larger than the length of this String object, or beginIndex is larger than endIndex. 

How to solve the StringIndexOutOfBoundsException

- Check the length of the string before using substring()
- Exception handling using try...catch.

**Comment by jcoglan** [comment]
@ilyash-b This should not be properly fixed. Whereas before you were seeing:

    if (offset < inputSize) {
        chunk0 = input.substring(offset, offset + 4);
    }

The code that's now on `master` will generate this, so that we only extract chunks if the upper bound is at or before the end of the input:

    int max0 = offset + 4;
    if (max0 <= inputSize) {
        chunk0 = input.substring(offset, max0);
    }

We don't have a unit test suite for every language; we have comprehensive tests for JavaScript, which the project is written in, that allow us to check fine details, and then we rely on the examples in the repo to check other languages. 

Ideally we'd have as little language-specific logic in each backend as possible, but unfortunately things like this slip through. The aim is to make sure the builder classes have robust ways of implementing all the required operations, and the rest of the compiler builds on top of that.

**Comment by jcoglan** [comment]
This should now be fixed as part of the v0.4 release; see https://blog.jcoglan.com/2022/04/01/announcing-canopy-0-4/
